File: apiserver/src/main/java/Metrics.java
/*
 * This file is part of Bob.
 *
 * Bob is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Bob is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Bob. If not, see <http://www.gnu.org/licenses/>.
 */

import crux.api.ICruxAPI;
import io.prometheus.client.CollectorRegistry;
import io.prometheus.client.Gauge;
import io.prometheus.client.exporter.common.TextFormat;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpServerRequest;
import io.vertx.ext.web.RoutingContext;
import io.vertx.rabbitmq.RabbitMQClient;

import java.io.IOException;
import java.io.Writer;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

public class Metrics {
    static final Gauge queuedEntities = Gauge
        .build()
        .name("bob_queued_entities")
        .help("Number of queued entity changes to be applied")
        .register();
    static final Gauge queuedJobs = Gauge
        .build()
        .name("bob_queued_jobs")
        .help("Number of queued jobs to be picked up")
        .register();
    static final Gauge errors = Gauge
        .build()
        .name("bob_errors")
        .help("Number of errors")
        .register();
    static final Gauge runningJobs = Gauge
        .build()
        .name("bob_running_jobs")
        .help("Number of jobs currently running")
        .register();
    static final Gauge failedJobs = Gauge
        .build()
        .name("bob_failed_jobs")
        .help("Number of failed jobs")
        .register();
    static final Gauge passedJobs = Gauge
        .build()
        .name("bob_passed_jobs")
        .help("Number of passed jobs")
        .register();
    static final Gauge pausedJobs = Gauge
        .build()
        .name("bob_paused_jobs")
        .help("Number of paused jobs")
        .register();
    static final Gauge stoppedJobs = Gauge
        .build()
        .name("bob_stopped_jobs")
        .help("Number of stopped jobs")
        .register();

    private final static CollectorRegistry registry = CollectorRegistry.defaultRegistry;

    private static class VertxBufferedWriter extends Writer {
        private final Buffer buffer = Buffer.buffer();

        @Override
        public void write(char[] buff, int off, int len) {
            buffer.appendString(new String(buff, off, len));
        }

        @Override
        public void flush() {
        }

        @Override
        public void close() {
        }

        Buffer getBuffer() {
            return buffer;
        }
    }

    private static Set<String> parse(HttpServerRequest request) {
        return new HashSet<>(request.params().getAll("name[]"));
    }

    private static void countJobs(ICruxAPI node) {
        Stream.of("running", "passed", "failed", "paused", "stopped")
            .parallel()
            .map(status -> {
                final var query = DB.datafy(
                    """
                    {:find  [run]
                     :where [[run :type :pipeline-run]
                             [run :status :%s]]}
                    """.formatted(status)
                );
                return Map.of(status, node.db().query(query).size());
            })
            .forEach(metric -> {
                final var status = metric.keySet().stream().findFirst().get();
                final var count = metric.get(status);

                switch (status) {
                    case "running" -> Metrics.runningJobs.set(count);
                    case "passed" -> Metrics.passedJobs.set(count);
                    case "failed" -> Metrics.failedJobs.set(count);
                    case "paused" -> Metrics.pausedJobs.set(count);
                    case "stopped" -> Metrics.stoppedJobs.set(count);
                }
            });
    }

    public static void writeMetrics(RoutingContext routingContext, RabbitMQClient queueClient, ICruxAPI node) {
        final var writer = new VertxBufferedWriter();

        queueClient
            .queueDeclare("bob.jobs", true, false, false)
            .compose(jobs -> {
                Metrics.queuedJobs.set(jobs.getMessageCount());
                return queueClient.queueDeclare("bob.entities", true, false, false);
            })
            .compose(entities -> {
                Metrics.queuedEntities.set(entities.getMessageCount());
                return queueClient.queueDeclare("bob.errors", true, false, false);
            })
            .onSuccess(errors -> {
                Metrics.errors.set(errors.getMessageCount());

                try {
                    countJobs(node);

                    TextFormat.write004(writer, registry.filteredMetricFamilySamples(parse(routingContext.request())));
                } catch (IOException e) {
                    Handlers.toJsonResponse(routingContext, e.getMessage(), 500);
                }
                routingContext.response()
                    .setStatusCode(200)
                    .putHeader("Content-Type", TextFormat.CONTENT_TYPE_004)
                    .end(writer.getBuffer());
            })
            .onFailure(err -> Handlers.toJsonResponse(routingContext, err.getMessage(), 500));
    }
}

File: apiserver-next/src/apiserver_next/metrics.clj
;   This file is part of Bob.
;
;   Bob is free software: you can redistribute it and/or modify
;   it under the terms of the GNU Affero General Public License as published by
;   the Free Software Foundation, either version 3 of the License, or
;   (at your option) any later version.
;
;   Bob is distributed in the hope that it will be useful,
;   but WITHOUT ANY WARRANTY; without even the implied warranty of
;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;   GNU Affero General Public License for more details.
;
;   You should have received a copy of the GNU Affero General Public License
;   along with Bob. If not, see <http://www.gnu.org/licenses/>.

(ns apiserver_next.metrics
  (:require [iapetos.core :as prometheus]
            [iapetos.export :as export]
            [failjure.core :as f]
            [crux.api :as crux]
            [langohr.queue :as lq]))

(defonce registry
         (-> (prometheus/collector-registry)
             (prometheus/register
               (prometheus/gauge :bob/queued-entities {:description "Number of queued entity changes to be applied"})
               (prometheus/gauge :bob/queued-jobs {:description "Number of queued jobs to be picked up"})
               (prometheus/gauge :bob/errors {:description "Number of errors"})
               (prometheus/gauge :bob/running-jobs {:description "Number of jobs currently running"})
               (prometheus/gauge :bob/failed-jobs {:description "Number of failed jobs"})
               (prometheus/gauge :bob/passed-jobs {:description "Number of passed jobs"})
               (prometheus/gauge :bob/paused-jobs {:description "Number of paused jobs"})
               (prometheus/gauge :bob/stopped-jobs {:description "Number of stopped jobs"}))))

(defn count-statuses
  [db]
  (f/try-all [statuses [:running :passed :failed :paused :stopped]
              counts   (pmap (fn [status]
                               {:status status
                                :count  (count (crux/q (crux/db db)
                                                       `{:find  [run]
                                                         :where [[run :type :pipeline-run]
                                                                 [run :status ~status]]}))})
                             statuses)]
    (doseq [{:keys [status count]} counts]
      (case status
        :running (prometheus/set (registry :bob/running-jobs) count)
        :passed  (prometheus/set (registry :bob/passed-jobs) count)
        :failed  (prometheus/set (registry :bob/failed-jobs) count)
        :paused  (prometheus/set (registry :bob/paused-jobs) count)
        :stopped (prometheus/set (registry :bob/stopped-jobs) count)))
    (f/when-failed [err]
      err)))

(defn collect-metrics
  [queue db]
  (f/try-all [_ (->> "bob.jobs"
                     (lq/message-count queue)
                     (prometheus/set (registry :bob/queued-jobs)))
              _ (->> "bob.entities"
                     (lq/message-count queue)
                     (prometheus/set (registry :bob/queued-entities)))
              _ (->> "bob.errors"
                     (lq/message-count queue)
                     (prometheus/set (registry :bob/errors)))
              _ (count-statuses db)]
    (export/text-format registry)
    (f/when-failed [err]
      err)))
